compiler/AST/iterator.cpp

```diff
@@ -67,6 +68,290 @@ IteratorInfo::IteratorInfo() :
+
+  // Use the formals of 'serial', for the purposes of resolution.
+  for_formals(formal, serial)
+    repCall->insertAtTail(
+      new NamedExpr(formal->name, createSymExprPropagatingParam(formal)) );
+
+  // Add the tag argument.
+  repCall->insertAtTail(new NamedExpr(astrTag, new SymExpr(iterKindTag)));
+
+  // Wrap in a block, to contain possible side-effects of resolving.
+  BlockStmt* container = new BlockStmt(repCall);
+  call->getStmtExpr()->insertAfter(container);
+
+  if (FnSymbol* parIter = tryResolveCall(repCall)) {
+    // Got it.
+    if (isIteratorOrForwarder(parIter)) {
+      igroupPtr = parIter;
+      parIter->iteratorGroup = igroup;
+
+    } else {
+      // If this is not an iterator, do not record it.
+      // We may need to raise an error later.
+      noniterFlag = true;
+    }
+  }
+
+  container->remove();
+}
+
+// See if any parallel iterators are available.
+void resolveAlsoParallelIterators(FnSymbol* serial, Expr* call) {
+  if (! isIteratorOrForwarder(serial)) return;  // not of interest
+  if (serial->iteratorGroup != NULL) return;  // already taken care of
+
+  // Similar to isIteratorOfType().
+  for_formals(formal, serial)
+    if (formal->name == astrTag && formal->type == gLeaderTag->type)
+      // 'serial' is actually a parallel iterator. Since we did not come
+      // from a serial iterator, we will not update iterator groups.
+      return;
+
+  IteratorGroup* igroup = new IteratorGroup();
+  igroup->serial = serial;
+  serial->iteratorGroup = igroup;
+
+  checkParallelIterator(serial, call, gStandaloneTag, igroup,
+                        igroup->standalone, igroup->noniterSA);
+
+  checkParallelIterator(serial, call, gLeaderTag, igroup,
+                        igroup->leader, igroup->noniterL);
+}
+
+static inline void verifyIGfunction(IteratorGroup* igroup, FnSymbol* fn) {
+  if (fn != NULL)
+    INT_ASSERT(fn->iteratorGroup == igroup);
+}
+
+void verifyIteratorGroup(FnSymbol* it) {
+  IteratorGroup* igroup = it->iteratorGroup;
+  if (! igroup) return;
+
+  INT_ASSERT(it == igroup->serial     ||
+             it == igroup->standalone ||
+             it == igroup->leader     ||
+             it == igroup->follower);
+
+  // Independent of 'it'.
+  INT_ASSERT(igroup->serial != NULL);
+  verifyIGfunction(igroup, igroup->serial);
+  verifyIGfunction(igroup, igroup->standalone);
+  verifyIGfunction(igroup, igroup->leader);
+  verifyIGfunction(igroup, igroup->follower);
+}
+
+static inline void cleanupIGfunction(FnSymbol* it, FnSymbol*& igroupPtr) {
+  if (it == igroupPtr) {
+    igroupPtr = NULL;
+    it->iteratorGroup = NULL;
+  }
+}
+
+// Actions upon deleting a FnSymbol.
+void cleanupIteratorGroup(FnSymbol* it) {
+  IteratorGroup* igroup = it->iteratorGroup;
+  if (! igroup) return;
+
+  it->iteratorGroup = NULL;
+
+  bool deleteIG = false;
+
+  if (it == igroup->serial) {
+    // Without the serial iterator, the group is of no interest.
```

Comment:
I find this a bit surprising but I'm OK with it for now. I'd expect function resolution to be able to produce resolved parallel iterators where the serial iterators are unused and cleaned up.

Reply:
> I'd expect function resolution to be able to produce resolved parallel iterators where the serial iterators are unused and cleaned up.

Not having a group in this case does not affect the ability to use a parallel iterator when there is no corresponding serial iterator. (If this is the concern here.) Regardless of whether the serial iterator was cleaned up or it did not exist at all.